integration.py

```diff
@@ -0,0 +1,256 @@
+
+        if report_path:
+            self.save(report_path, test_results)
+
+
+class Environment(object):
+    @staticmethod
+    def checkout(rev):
+        run(["git", "submodule", "foreach", "git", "checkout", rev])
+
+    @staticmethod
+    def sync():
+        run(["bash", "sync.sh"])
+
+
+def diff(json_a_path, json_b_path, output_path):
+    # extract a subset
+    def _transform(path):
+        """The following jq expression can be used for document comparison
+
+            jq \
+                '.results | to_entries |
+                map({doc_type: .key, error_rate: .value.error_rate})' \
+            test-reports/dev.report.json
+        """
+        with open(path, 'r') as f:
+            data = json.loads(f.read())
+        subset = {}
+        # iterate over the measurements and collect comparable stats
+        for doc_type, measures in data['results'].items():
+            subset[doc_type] = {
+                'error_rate': measures['error_rate']
+            }
+        return json.dumps(subset, indent=4).splitlines(keepends=True)
+
+    json_a = _transform(json_a_path)
+    json_b = _transform(json_b_path)
+
+    result = difflib.unified_diff(json_a, json_b)
+    output = ''.join(result)
+    print(output)
+
+    print("Writing diff to {}".format(output_path))
+    with open(output_path, 'w') as f:
+        f.write(output)
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.command()
+@click.option('--data-path', type=click.Path(exists=True),
+              default='resources/data')
+@click.option('--report-path', type=click.Path(dir_okay=False))
+def report(data_path, report_path):
+    """Run an integration report against currently loaded schemas."""
+    Reporter().run(data_path, report_path)
+
+
+@cli.command()
+@click.argument('rev-A')
+@click.argument('rev-B')
+@click.option('--data-path', type=click.Path(), default='resources/data')
+@click.option('--report-path', type=click.Path(file_okay=False), required=True)
+@click.option('--cache/--no-cache', default=True)
+def compare(rev_a, rev_b, data_path, report_path, cache):
+    """Compare the results of two revisions of `mozilla-pipeline-schemas`."""
+
+    if os.environ.get("EXTERNAL"):
+        err_msg = "EXTERNAL configuration is currently not supported"
+        raise NotImplementedError(err_msg)
+
+    def _run_report(rev):
+        output_path = os.path.join(report_path, "{}.report.json".format(rev))
+
+        # exit early if the report already been run and we are using the cache
+        if os.path.exists(output_path) and cache:
+            return output_path
+
+        Environment.checkout(rev)
+        Environment.sync()
+        Reporter().run(data_path, output_path)
+
+        return output_path
+
+    rev_a_path = _run_report(rev_a)
+    rev_b_path = _run_report(rev_b)
+
+    diff_path = os.path.join(report_path, "{}-{}.diff".format(rev_a, rev_b))
+    diff(rev_a_path, rev_b_path, diff_path)
+
+    # TODO: restore the proper state
```

Comment:
Hm.

Reply:
It requires parsing stdout for `git rev-parse HEAD` and restoring it after. Restoring to dev is not a terrible assumption at this point.